### ISSUE TITLE
Release 5.2.1

### ISSUE THREAD

**Issue by stakx** [castleproject/Core#662]
I think we're almost ready for another release.

 * The PRs #664 and #665 are both marked as drafts, and will require some more work. They probably won't be finalized it time for this release.
 * PR #661 is not a draft, and I'm hoping we can include it here. But if it needs more work & time, we could also do a release 5.1.2 without it instead.

----

### Things to do _after_ this next release

I'd love if we could tackle those larger discussions about this library's (and DynamicProxy's) future, possibly resulting in a major version bump:

 * whether or not (and how) to extract deprecated DictionaryAdapter into a separate project (#394);
 * whether or not to abandon (especially binary) serialization of DynamicProxy proxies (#562);
 * what to do about the logging abstractions & integration packages (#408).

Depending on the outcome of those decisions, we may end up with a much slimmer library and a somewhat pared-down DynamicProxy implementation that's easier to manage, so I think it would be good to start there and get rid of as much "baggage" as soon as we can.

(If we decide to keep the logging abstractions, we could take a look at #418, which should be relatively straightforward.)

I'd also love to do some code exploration about whether we really need 5 different types of proxy; I suspect we have a _lot_ of code duplication in DynamicProxy because of that. While working on #447, that suspicion has only increased, since one of the previously distinguishing characteristics of interfaces &ndash; that they may not contain any implementation bits &ndash; no longer holds true in modern C# and .NET. The distinction between class and interface proxies may no longer make sense. (Note that instead of `CreateClassProxy<TClass>()`, you can already do an practically equivalent `(TClass)CreateInterfaceProxyWithoutTarget<IEmptyInterface>(new ProxyGenerationOptions { BaseTypeForInterfaceProxy = typeof(TClass) })`). We could potentially cut the number of collectors / contributors / generators in half with only a few minor breaking changes, and simplify the public API of `ProxyGenerator` a lot.

**Comment by stakx** [comment]
I've been delayed with daytime (aka paid) work, but I think I'll find some time very soon to finally get DIM support published.

**Comment by stakx** [comment]
Apologies to everyone who's been awaiting this release. Version 5.2.0 is basically ready, as soon as we've dealt with an expired NuGet API key, we should be able to publish it to NuGet.

**Comment by 304NotModified** [comment]
This is still open, isn't?

**Comment by stakx** [comment]
@304NotModified, sorry for the delay. I *could* release 5.2.0 (and I have wanted to for a long time), but there is #684. I am somewhat worried about breaking e.g. NSubstitute by releasing 5.2.0 as is. Currently very busy with work so I don't have much time to prepare a fix in advance.

**Comment by stakx** [comment]
I just tried to re-release this as 5.2.1 only for the CI build to fail immediately. We need to fix that first (related: #683).

**Comment by 304NotModified** [comment]
I see 5.2.1 now on nuget.org 🎉

**Comment by stakx** [comment]
You're right. I got so many pipeline failure emails that I overlooked the one that mattered re: publication didn't fail. It's a bit of a mess but I'm glad 5.2.1 made it through the door. The CI builds will still need fixing but we can do that for the next release.

**Comment by 304NotModified** [comment]
Thanks @stakx!

Can you update the release on GitHub? That will prevent confusion in the future.

https://github.com/castleproject/Core/releases

**Comment by stakx** [comment]
> Can you update the release/tag on GitHub? That will prevent confusion in the future. 

@304NotModified, thanks for the hint. I trashed the 5.2.0 release (which still carried the unmet August 2024 release date) and recreated a 5.2.1 one.

As you can perhaps tell, I've probably spent too much time away from this project, I feel a little disorganized right now.

**Comment by stakx** [comment]
Also @304NotModified:

> there are some build errors with NSubstitute [...] This isn't the same as https://github.com/castleproject/Core/issues/684 right?

No, those build errors seem unrelated to #684. I'd say they're a direct result of #668.

**Comment by 304NotModified** [comment]
@stakx could you please create a new release? :)

**Comment by stakx** [comment]
@304NotModified, I'll look into it.

* I'm assuming you're thinking primarily about #684? Have you verified that you can get your side working properly after our bugfix for this?

* Would you require anything else to go into a patch release? I've done a fair bit of code refactoring, so currently not sure what else could easily be backported.

**Comment by 304NotModified** [comment]
> * I'm assuming you're thinking primarily about [`invocation.MethodInvocationTarget` throws `ArgumentNullException` for default interface method #684](https://github.com/castleproject/Core/issues/684)? 

Yes indeed :)

> * Have you verified that you can get your side working properly after our bugfix for this

If there is a test package, I would be happy to test it.👍

**Comment by 304NotModified** [comment]
I will try to the test this next week.

**Comment by 304NotModified** [comment]
@stakx I can confirm that this issue in Castle Core 5.2.1 has been fixed in 6.0.0 preview (taken from https://ci.appveyor.com/project/castleproject/core/builds/53542581/job/b0vpahb8rymgngak/artifacts)
